### PR TITLE
[DataLoader] Add shuffling logic for MapDataPipe

### PR DESCRIPTION
**Pull request by ejguan**
Fixes: https://github.com/pytorch/data/issues/718

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#82975 [DataLoader] Add shuffling logic for MapDataPipe**
* #83202 [DataPipe] Convert MapDataPipe.shuffle to IterDataPipe

## Changes:
- Add shuffling logic for the graph of `DataPipe` with `MapDataPipe` at the end when using `DataLoader`
  - Apply `shuffle` (argument from `DataLoader`)
  - Assign new shuffle seed to each random `DataPipe` per iteration
- When it's a `MapDataPipe` at the end of graph, we only need to run distributed sharding without counting the number of workers because index sampler will do the local sharding work.

### CODE REVIEW

**Review by NivekT — APPROVED** [review]
This LGTM! We should make sure at least one of the PRs have tests with `ShufflerMapDataPipe` used with `DataLoader` (original version).

File: test/test_dataloader.py [hunk 1]
```diff
@@ -2262,10 +2276,31 @@ def test_shuffle(self):
...
+
+        dl = DataLoader2(dp, batch_size=None, num_workers=2, shuffle=True)
+        self.assertNotEqual(items, list(dl))
+        self.assertEqual(items, sorted(list(dl)))
+
+        dl = DataLoader2(self.MapSorter(dp), batch_size=None, num_workers=2, shuffle=True)
         self.assertEqual(list(dl), items)
 
-        dl = DataLoader2(self.Sorter(dp), batch_size=None, num_workers=2, shuffle=True)
+        dl = DataLoader2(self.MapSorter(dp), batch_size=None, num_workers=2, shuffle=True)
```
Comment (ejguan) [inline test/test_dataloader.py:2303]: I can convert the `DataLoader2` to `DataLoader` to make it clear. This `DataLoader2` is actually used `DataLoader` internally: https://github.com/pytorch/pytorch/blob/9c5c045f63fc142a4d9e6e132912236f1dd2bc7d/torch/utils/data/dataloader_experimental.py#L99-L111

cc: @NivekT

Comment (NivekT) [reply test/test_dataloader.py:2303]: Let's do that since eventually they will be different (or we will remove the experimental from here).

File: test/test_dataloader.py [hunk 2]
```diff
@@ -2262,10 +2276,31 @@ def test_shuffle(self):
...
+
+        dl = DataLoader2(self.IterSorter(dp), batch_size=None, num_workers=2, shuffle=True)
+        self.assertEqual(list(dl), items)
+
+    def test_map_shuffle(self):
+        items = list(range(1000))
+        dp = SequenceWrapper(items).shuffle()
+
+        dl = DataLoader2(dp, batch_size=None, num_workers=2, shuffle=False)
+        self.assertEqual(items, list(dl))
```
Comment (BarclayII) [inline test/test_dataloader.py:2290]: I guess this PR is not to handle the difference between `IterDataPipe.shuffle()` (whose order changes at every epoch) and `MapDataPipe.shuffle()` (whose order do not change)?

Comment (ejguan) [reply test/test_dataloader.py:2290]: This PR guarantees those `DataPipes` are shuffled differently by `DataLoader` per epoch

File: torch/utils/data/dataloader.py [hunk 3]
```diff
@@ -105,10 +105,14 @@ def _get_distributed_settings():
...
-    global_worker_id = worker_id
+def _sharding_worker_init_fn(is_iter, worker_init_fn, world_size, rank_id, worker_id):
     info = torch.utils.data.get_worker_info()
-    total_workers = info.num_workers
+    if is_iter:
+        global_worker_id = worker_id
+        total_workers = info.num_workers
+    else:
+        global_worker_id = 0
+        total_workers = 1
```
Comment (ejguan) [inline torch/utils/data/dataloader.py:115]: We only need to apply distributed + mp sharding for `IterDataPipe`. For `MapDataPipe`, it seems distributed sharding is the only requirement.

File: test/test_dataloader.py (hunk 2 above)
Comment (BarclayII) [reply test/test_dataloader.py:2292]: Then does that mean `shuffle()` in both IterDataPipe and MapDataPipe will have no effect if `shuffle` argument in `DataLoader` is False?  If so, then why do we still need the `shuffle()` DataPipe?

Comment (ejguan) [reply test/test_dataloader.py:2292]: This is because users would be able to choose either shuffle or not with the same pipeline.

Comment (BarclayII) [reply test/test_dataloader.py:2292]: I see.  Thanks for the clarification!

File: torch/utils/data/dataloader.py (hunk 3 above)
Comment (NivekT) [reply torch/utils/data/dataloader.py:115]: It seems like it is possible that users start off with `IterDataPIpe`, add a `sharding_filer`, then convert to `MapDataPipe` prior to passing it into `DataLoader`.

IIUC this might not handle that case properly? We can potentially prevent that case with a linter?

**Review by NivekT — COMMENTED** [review]
LGTM but see comment. Are mixed usages of `IterDataPipe` and `MapDataPipe` edge cases that we need to worry about?

File: torch/utils/data/dataloader.py (hunk 3 above)
Comment (ejguan) [reply torch/utils/data/dataloader.py:115]: It should be fine. And, it's a great idea to add a test to guard the behavior. Let me do that

Comment (NivekT) [reply torch/utils/data/dataloader.py:115]: Will sharding still be correctly applied if the last DataPipe that user provides is `to_map_datapipe`?

Comment (ejguan) [reply torch/utils/data/dataloader.py:115]: I think so because `DataLoader` would shard the indices in the main process. So, as long as the end DataPipe is a `MapDataPipe`, the whole pipeline should be sharded based on number of workers.

I might not be able to add a test here due to `to_map_datapipe` is in TorchData. But, I will definitely verify the behavior locally.

**Review by NivekT — APPROVED** [review]

File: torch/utils/data/dataloader.py (hunk 3 above)
Comment (ejguan) [reply torch/utils/data/dataloader.py:115]: It turns out `shuffle` is way more complicated than I though. I might need to fully revamp the logic of this PR. Need to do more investigation!

Comment (ejguan) [reply torch/utils/data/dataloader.py:115]: BTW, it's also blocked by this issue: https://github.com/pytorch/data/issues/753
